File: 3604_Minimum_Time_to_Reach_Destination_in_Directed_Graph.py
import heapq
from typing import List, Tuple

class Solution:
    def minTime(self, n: int, edges: List[List[int]]) -> int:
        adj: List[List[Tuple[int, Tuple[int,int]]]] = [[] for _ in range(n)]
        for u, v, l, r in edges:
            adj[u].append((v, (l, r)))
        
        dist = [float('inf')] * n
        pq = []
        heapq.heappush(pq, (0, 0))
        dist[0] = 0
        
        while pq:
            time, node = heapq.heappop(pq)
            
            if node == n - 1:
                return time
            if time > dist[node]:
                continue
            
            for neigh, (l, r) in adj[node]:
                if time > r:
                    continue
                
                curr_time = max(time, l)
                
                if dist[neigh] > curr_time + 1:
                    dist[neigh] = curr_time + 1
                    heapq.heappush(pq, (curr_time + 1, neigh))
        
        return -1

File: cpp/3604_Minimum_Time_to_Reach_Destination_in_Directed_Graph.cpp
class Solution {
public:
    int minTime(int n, vector<vector<int>>& edges) {
        vector<vector<pair<int, pair<int, int>>>> adj(n);

        for (auto& e : edges)
            adj[e[0]].push_back({e[1], {e[2], e[3]}});

        vector<int> dist(n, INT_MAX);
        priority_queue<pair<int, int>, vector<pair<int, int>>, greater<>> pq;

        pq.push({0, 0});
        dist[0] = 0;

        while (!pq.empty()) 
        {
            auto [time, node] = pq.top();
            pq.pop();

            if (node == n - 1) return time;
            if (time > dist[node]) continue;

            for (auto& [neigh, range] : adj[node])
            {
                if (time > range.second)
                    continue;

                int curr_time = max(time, range.first);

                if (dist[neigh] > curr_time + 1) 
                {
                    dist[neigh] = curr_time + 1;
                    pq.push({curr_time + 1, neigh});
                }
            }
        }

        return -1;
    }
};

